pkg/server/chaosd/disk.go

```diff
@@ -38,84 +38,118 @@ const DDWritePayloadCommand = "dd if=/dev/zero of=%s bs=%s count=%s oflag=dsync"
+	attack := options.(*core.DiskOption)
 
 	if options.String() == core.DiskFillAction {
 		return disk.attackDiskFill(attack)
 	}
 	return disk.attackDiskPayload(attack)
 }
 
-func (diskAttack) attackDiskPayload(payload *core.DiskCommand) error {
+func (diskAttack) attackDiskPayload(payload *core.DiskOption) error {
+	var cmdFormat string
 	switch payload.Action {
 	case core.DiskWritePayloadAction:
+		cmdFormat = DDWritePayloadCommand
 		if payload.Path == "" {
-			payload.Path = "/dev/null"
-		}
-		cmd := exec.Command("bash", "-c", fmt.Sprintf(DDWritePayloadCommand, payload.Path, "1M", payload.Size))
-		output, err := cmd.CombinedOutput()
-
-		if err != nil {
-			log.Error(string(output), zap.Error(err))
-		} else {
-			log.Info(string(output))
+			var err error
+			payload.Path, err = utils.CreateTempFile()
+			if err != nil {
+				log.Error(fmt.Sprintf("unexpected err when CreateTempFile in action: %s", payload.Action))
+				return err
+			}
+			defer func() {
+				err := os.Remove(payload.Path)
+				if err != nil {
+					log.Error(fmt.Sprintf("unexpected err when removing temp file %s", payload.Path), zap.Error(err))
+				}
+			}()
 		}
-		return err
 	case core.DiskReadPayloadAction:
+		cmdFormat = DDReadPayloadCommand
 		if payload.Path == "" {
-			err := errors.Errorf("empty read payload path")
-			log.Error(fmt.Sprintf("payload action: %s", payload.Action), zap.Error(err))
-			return err
-		}
-		cmd := exec.Command("bash", "-c", fmt.Sprintf(DDReadPayloadCommand, payload.Path, "1M", payload.Size))
-		output, err := cmd.CombinedOutput()
-
-		if err != nil {
-			log.Error(string(output), zap.Error(err))
-		} else {
-			log.Info(string(output))
+			path, err := utils.GetRootDevice()
+			if err != nil {
+				log.Error("err when GetRootDevice in reading payload", zap.Error(err))
+				return err
+			}
+			if path == "" {
+				err = errors.Errorf("can not get root device path")
+				log.Error(fmt.Sprintf("payload action: %s", payload.Action), zap.Error(err))
+				return err
+			}
+			payload.Path = path
 		}
-		return err
 	default:
 		err := errors.Errorf("invalid payload action")
 		log.Error(fmt.Sprintf("payload action: %s", payload.Action), zap.Error(err))
 		return err
 	}
+
+	byteSize, err := utils.ParseUnit(payload.Size)
+	if err != nil {
+		log.Error(fmt.Sprintf("fail to get parse size per units , %s", payload.Size), zap.Error(err))
+		return err
+	}
+	sizeBlocks := utils.SplitByteSize(byteSize, payload.PayloadProcessNum)
+
+	var wg sync.WaitGroup
+	wg.Add(len(sizeBlocks))
+	fatalErrors := make(chan error)
+	wgDone := make(chan bool)
+	for _, sizeBlock := range sizeBlocks {
+		cmd := exec.Command("bash", "-c", fmt.Sprintf(cmdFormat, payload.Path, sizeBlock.BlockSize, sizeBlock.Size))
+
+		go func(cmd *exec.Cmd) {
+			output, err := cmd.CombinedOutput()
+
+			if err != nil {
+				log.Error(string(output), zap.Error(err))
+				fatalErrors <- err
+			}
+			log.Info(string(output))
+			wg.Done()
+		}(cmd)
+	}
+
+	go func() {
+		wg.Wait()
```

Comment:
why use `wait` in goroutine?

Reply:
wg.Wait() will wait blocks until the wg counter is zero. wg.Add(len(sizeBlocks)) will add len(sizeBlocks) to wg, and wg.Done is wg.Add(-1).